README.md
# README #

# Usage:
Start NaviServer with the sample config in one window in the foreground:

````
$ /usr/local/ns/bin/nsd -t nscoap-config.tcl -f
````

The sample config file maps GET and POST requests to /foo/* for
forwarding to the HTTP server. The sample config file registers as
well a request handler for POST and GET on /foo/bar via
`ns_register_proc`. One can certainly run other requests by adding
tcl/adp/html files into the pages directory, or via other means
possible by NaviServer.

To issue a CoAP GET request from other terminal, one can use e.g. the
wget like coap-client from https://github.com/obgm/libcoap.git:

````
$ coap-client -m get coap://localhost/foo/bar
GET-okidoki
````

Issue a CoAP POST request (assuming, you have
a file called TODO in your current directory):

````
$ coap-client -m post -f TODO coap://localhost/foo/bar
POST-okidoki(961)
````

The number between parens is the number of bytes transferred in
body of the POST request (check the registered proc in the
sample config file `nscoap-config.tcl`.

When issuing these requests, the terminal running the server in the
foreground shows some debugging information prefixed with
`Debug(coap)`.  This output can be controlled e.g. in the config file
via `ns_logctl severity Debug(coap) on|off`.


Changelog
commit 8ac115fbe677d5d051e0dc00ff3b19f6b34288e9
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2023-07-13

    Added .gitignore

M	.gitignore

commit 11560aa4bca7b5d56d7acebf9d2018f03f9c3bc9
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2023-07-13

    Maintenance work:
    
    - Reduced verbosity (don't say "loaded" for every interpreter init)
    - Improved spelling

M	README.md
M	nscoap-config.tcl
M	nscoap.c

commit 4c24ca15b997a9183a4bffb41f0e74c685f30de2
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2023-04-19

    avoid compiler warning when compiled against Tcl9

M	nscoap.c

commit 10e8dd944134c819c9a88a9dff3ea43f82b67ecd
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2022-06-13

    added Changelog

M	.gitignore

commit 1ef7b1ee7cbe7e1d4ed018a45b8fdc2207e5cc3c
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2022-02-05

    improve spelling

M	nscoap-config.tcl
M	nscoap.c
M	nscoap.h

commit 07e9568a06dc5443c4018ff353fa60459b45b6f6
Merge: 6634a83 a4dc532
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2021-02-12

    Merge branch 'master' of https://bitbucket.org/naviserver/nscoap

commit a4dc532237a5086238036441d0a6fd9f4bf44bc9
Author: Gustaf Neumann <neumann@wu.ac.at>
Date:   2021-02-12

    Initial commit

A	.gitignore

commit 6634a83a281a11b8497678206413d812c5d5e7d9
Author: David Hopfmueller <david@hopfmueller.at>
Date:   2020-07-11

    Remove outdated usage instructions

M	README.md

commit bdcea73b8704683b0d393447806eaa8f5b700134
Author: David Hopfmueller <david@hopfmueller.at>
Date:   2020-07-11

    Bump NS_VERSION to 0.2

M	nscoap.h

commit 88c071de47a48e3843c6fe119675009d707ab08d
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2019-06-07

    minor code cleanup:
    - reduce reference chains
    - remove redundant assignments
    - remove locical error

M	nscoap.c

commit c6cafbf009afea286569e552ad176d20840be188
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2019-06-06

    added content types for POST/PUT requests as defined in RFC 7252

M	nscoap.c
M	nscoap.h

commit 3129969255e8c7f57c9bca08a8fd8076e6eb2103
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2019-06-06

    don't hardcode get on the Ns_UrlSpecificGet() call

M	nscoap.c

commit ec7d1ecbf6bab46223ee865dbbeca1f9c6cb0db5
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2019-06-05

    addin examples to the README.md file

M	README.md
M	nscoap-config.tcl

commit f2f444edde3ee30df3d678b0097c0ef709db702c
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2019-06-05

    whitespace changes, improve spelling

M	nscoap.c
M	tests/nscoap.test

commit 4db7838cd55f6346cff6739c5890ccd6e0aecb28
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2019-06-05

    - add payload handling for CoAP POST requests
    - fix path construction
    - remove redundant operations
    - add mapping example in sample config file
    - turn on coap debug for now

M	nscoap-config.tcl
M	nscoap.c
M	nscoap.h

commit fa76dd0edd4cf2177f418ee621e0da2a7cb2449a
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2018-08-05

    - add comment how to set values in the nsv (see end of config file)
    - add more debugging aids when Debug(coap) is on

M	nscoap-config.tcl
M	nscoap.c
M	tests/config.tcl

commit 34c9caae43dcee0942fd77631322acc5a12e80b9
Author: David Hopfmueller <david.hopfmueller@wu.ac.at>
Date:   2017-12-14

    Add additional tests

M	tests/nscoap.test

commit 6096563704162450a0bc7bcd00062e54f7dcaa1a
Author: David Hopfmueller <david.hopfmueller@wu.ac.at>
Date:   2017-12-13

    Set status code to prevent reuse of the previous request's value

M	nscoap.c

commit 86e968568aa4357e3e4fafc91af83cc3573445d4
Author: David Hopfmueller <david.hopfmueller@wu.ac.at>
Date:   2017-12-13

    Rework NSV logic
    
    - Use option number, restore option delta semantics
    - Use "nsv" instead of "nomap" keyword
    - Improve NSV key handling: the previous logic could result in segfaults as option values can indeed terminate CoAP packets
    - In order to use NSV handling, use an URI of /nsv/<key> where <key> is a string of 1..12 characters length
    - Fix bug when NSV key wasn't foung
    - Adaption of tests to reflect the new NSV logic

M	nscoap.c
M	tests/nscoap.test

commit a9a7fa48d7378f4fc64d6685d669362d409eaf3f
Author: David Hopfmueller <david.hopfmueller@wu.ac.at>
Date:   2017-12-13

    Track option number in addition to delta
    
    The number property is the result of the current option's delta + the previous option's number.  This retains the delta value which can be used to detect multiple occurrences of the same option type.

M	nscoap.h

commit d4152ddcbed0b92dd1bc965fb7862646f71ce2a2
Author: David Hopfmueller <david.hopfmueller@wu.ac.at>
Date:   2017-12-13

    Support individual and debugger testing targets

M	Makefile

commit e16f0f6d60a0ba90940cbc2256ff7dfae58a78d1
Author: David Hopfmueller <david.hopfmueller@wu.ac.at>
Date:   2017-12-11

    Add vim modeline

M	tests/nscoap.test

commit 58bc64a87bc1493174cec9be81500def00ddcc87
Author: David Hopfmueller <david.hopfmueller@wu.ac.at>
Date:   2017-12-11

    Make special handling modes standards-compliant
    
    When no resource is found, this should result in a 404. Direct reply for the special case of no URI path option.

M	nscoap.c

commit 97042ed7d9135c6f12719042c887bf45e1c8e30c
Author: David Hopfmueller <david.hopfmueller@wu.ac.at>
Date:   2017-12-11

    Make HTTP the default mapping method
    
    The previously used '/' is not ideal for matching since it usually isn't part of the URI path option value.

M	nscoap.c

commit 6a3a5b99fba02cf29757f718fc4b11f6ae0ac90d
Author: David Hopfmueller <david.hopfmueller@wu.ac.at>
Date:   2017-12-11

    Add vim modeline (NS project uses spaces)

M	nscoap.c

commit 448674606417c8e37355c3d0e6c7cd6a87148b5f
Author: David Hopfmueller <david.hopfmueller@wu.ac.at>
Date:   2017-12-11

    Don't expect an URI path option to be the first

M	nscoap.c


tests/all.tcl
#
# The contents of this file are subject to the Mozilla Public License
# Version 1.1 (the "License"); you may not use this file except in
# compliance with the License. You may obtain a copy of the License at
# http://mozilla.org/
#
# Software distributed under the License is distributed on an "AS IS"
# basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See
# the License for the specific language governing rights and limitations
# under the License.
#
# Copyright 2006 (C) Stephen Deasey <sdeasey@gmail.com>
#
# Alternatively, the contents of this file may be used under the terms
# of the GNU General Public License (the "GPL"), in which case the
# provisions of GPL are applicable instead of those above.  If you wish
# to allow use of your version of this file only under the terms of the
# GPL and not to allow others to use your version of this file under the
# License, indicate your decision by deleting the provisions above and
# replace them with the notice and other provisions required by the GPL.
# If you do not delete the provisions above, a recipient may use your
# version of this file under either the License or the GPL.
#

#
# all.tcl --
#
#       This file contains a top-level script to run all of the tests.
#       Execute it by invoking "source all.tcl" when running nsd in
#       command mode in this directory.
#

package require Tcl 8.5
package require tcltest 2.2
namespace import tcltest::*

configure {*}$argv -singleproc true -testdir [file dirname [info script]]

runAllTests


tests/config.tcl
set homedir [pwd]/tests
set bindir  [file dirname [ns_info nsd]]

ns_section ns/parameters
ns_param   home             $homedir
ns_param   tcllibrary       $bindir/../tcl

ns_section "ns/servers"
ns_param   test            "Test COAP Server"

ns_section ns/server/test/tcl
ns_param   initfile         $bindir/init.tcl

ns_section ns/server/test/modules
ns_param   nscoap           [pwd]/nscoap.so

ns_section   ns/server/test
ns_param     connsperthread 100000  ;# default: 0; number of connections (requests) handled per thread
ns_param     minthreads     2       ;# default: 1; minimal number of connection threads
ns_param     maxthreads     100     ;# default: 10; maximal number of connection threads
ns_param     maxconnections 100     ;# default: 100; number of allocated connection structures

ns_section   ns/server/test/module/nscoap
ns_param     address        localhost
ns_param     port           5683
ns_param     mapHTTP       "GET /*"
#ns_param    driverthreads  10          ;# default: 1
#ns_param    acceptsize      1          ;# default: value of "backlog"; max number of accepted (but unqueued) connection requests

# For debugging, you might activate one of the following flags
#
#ns_logctl severity Debug(coap) on
#ns_logctl severity Debug(ns:driver) on
#ns_logctl severity Debug(request) on


nscoap.h
/* Include NS declarations */
#include "ns.h"

/* Version of this plugin */
#define NSCOAP_VERSION "0.2"

/* Control debug output */
#define DEBUG 1

/* Maximum size (bytes) of a CoAP message as suggested by RFC 7252 */
#define MAX_COAP_SIZE 1152

/* Maximum size of a packet that's read from file. */
#define MAX_PACKET_SIZE 16384

/* Fallback value as per RFC 7252 section 4.6 */
#define MAX_COAP_CONTENT 1024

/* HTTP version to be used in requests */
#define HTTP_VERSION "HTTP/1.0"

typedef unsigned char byte;

/* Message code details */
typedef struct Code_s
{
    unsigned int     class;
    unsigned int     detail;
} Code_t;

typedef struct Option_s
{
    unsigned int delta;
    unsigned int length;
    unsigned int number;
    byte        *value;
} Option_t;

/* Holds the CoAP message and its parameters */
typedef struct CoapMsg_s
{
    bool         valid;          /* validity of the message */
    unsigned int version;        /* CoAP version */
    unsigned int type;           /* message type */
    size_t       tokenLength;    /* token length */
    unsigned int codeValue;      /* message code */
    unsigned int messageID;      /* message id */
    byte         token[8];       /* token */
    unsigned int contentFormat;  /* Content-Format RFC 7252 sect 5.10 */
    byte        *payload;        /* payload */
    size_t       payloadLength;  /* length of payload */
    int          optionCount;    /* number of valid options */
    Option_t    *options[MAX_COAP_CONTENT];
} CoapMsg_t;

typedef struct HttpReq_s
{
    const char *method;         /* HTTP method code */
    Ns_DString  token;          /* CoAP token */
    Ns_DString  host;           /* CoAP/HTTP URI host portion  */
    Ns_DString  path;           /* CoAP/HTTP URI path portion  */
    Ns_DString  query;          /* CoAP/HTTP URI query portion */
    const char *contentType;    /* HTTP Content-Type */
    byte       *payload;        /* pointer to beginning of payload */
    size_t      payloadLength;  /* length of payload in bytes */
} HttpReq_t;

typedef struct HttpRep_s
{
    unsigned int  status;                 /* HTTP status code */
    Ns_Set       *headers;                /* HTTP headers */
    byte         *payload;                /* pointer to beginning of payload */
    size_t        payloadLength;          /* length of payload in bytes */
} HttpRep_t;

/* Raw packet bytes (L4 payload) */
typedef struct Packet_s
{
    byte       *raw;                    /* holds the raw packet bytes */
    size_t      position;               /* current parser position */
    size_t      size;                   /* size in bytes */
} Packet_t;

/* Driver parameters */
typedef struct {
    int placeholder;                    /* currently, we need no "own" parameters,
                                           but empty structs are a non-standard an GNU extension */
} CoapDriver;

/* CoAP parameters (sock->arg) */
typedef struct CoapParams_s
{
    Ns_DString  *sendbuf;                /* buffered response waiting to be sent */
    unsigned int type;                   /* message type (for con/ack matching) */
    unsigned int messageID;              /* message ID (for req/ack matching) */
    byte         token[8];               /* CoAP token */
    size_t       tokenLength;            /* token length */
    unsigned int flags;
} CoapParams_t;

#define NSCOAP_FLAG_ALREADY_HANDLED 0x01u

/* Locally defined functions */
static bool SerializeCoap(CoapMsg_t *coap, Packet_t *packet);
static bool ParseCoap(Packet_t *packet, CoapMsg_t *coap, CoapParams_t *params);
static bool Coap2Http(CoapMsg_t *coap, HttpReq_t *http);
static bool Http2Coap(HttpRep_t *http, CoapMsg_t *coap, CoapParams_t *params);

static Ns_DriverListenProc Listen;
static Ns_DriverAcceptProc Accept;
static Ns_DriverRecvProc Recv;
static Ns_DriverSendProc Send;
static Ns_DriverKeepProc Keep;
static Ns_DriverCloseProc Close;
static Ns_TclTraceProc CoapInterpInit;
static Tcl_ObjCmdProc CoapObjCmd;

/*
 * Local Variables:
 * mode: c
 * c-basic-offset: 4
 * fill-column: 78
 * indent-tabs-mode: nil
 * End:
 */


coap2http.c
/*
 * CoAP to HTTP converter, standalone version
 */

#include "coap2http.h"

int main(int argc, char *argv[])
{
    CoapMsg_t request;

    bzero(&request, sizeof(CoapMsg_t));
    request.valid = NS_TRUE;

    if (argc != 2) {
        fprintf(stderr, "Usage: coap2http <captured.file>\n");
        exit(-1);
    }
    if (LoadMessage(argv[1], &request) == NS_TRUE) {
        fprintf(stderr, "Result: %d\n", ParseCoapMessage(&request));
    }

    return 0;
}

/*
 * Load message from file and save it to buffer together with its size
 */
static bool LoadMessage(char *file, CoapMsg_t *request)
{
    FILE *fd;
    bool success = NS_TRUE;

    if ((fd = fopen(file, "r"))) {
        request->size = (int) fread(request->raw, 1, MAX_SIZE, fd);
        fclose(fd);
    } else {
        fprintf(stderr, "File <%s> could not be opened.\n", file);
        success = NS_FALSE;
    }

    return success;
}

/*
 * Check size of the CoAP message
 */
static bool CheckRemainingSize(CoapMsg_t *request, int increment)
{
    bool success = NS_TRUE;

    if ((request->position + increment) > request->size) {
        request->valid = NS_FALSE;
        success = NS_FALSE;
    }

    return success;
}

/*
 * Parse the content of a CoAP request
 */
static bool ParseCoapMessage(CoapMsg_t *request) {
    Option_t option;
    int      i, codeValue, lastOptionNumber = 0;
    bool     processOptions;
    Code_t   code;
    /*
     * Registry of valid CoAP message codes
     */
    static const int code_registry[] = {
        000, 001, 002, 003, 004,
        201, 202, 203, 204, 205,
        400, 401, 402, 403, 404, 405, 406, 412, 413, 415,
        500, 501, 502, 503, 504, 505
    };

    request->position = 0;
    bzero(&option, sizeof(Option_t));
    request->valid = NS_TRUE;

#ifdef DEBUG
    fprintf(stderr, "parseRequest started.\n");
#endif
    /*
     * CoAP messages can't be shorter than 4 bytes
     */
    if (CheckRemainingSize(request, 4) == NS_FALSE) {
#ifdef DEBUG
        fprintf(stderr, "Message shorter than 4 bytes.\n");
#endif
        return NS_FALSE;
    }

    /*
     * Bit 0-1: CoAP version: Must be 0b01
     */
    request->version = ((request->raw[0] >> 6) == 0x1u);
#ifdef DEBUG
    fprintf(stderr, "CoAP version: %d\n", request->version);
#endif
    if (request->version == 0) {
        request->valid = NS_FALSE;
        return NS_FALSE;
    }

    /*
     * Bit 2-3: Type of CoAP message
     */
    request->type = ((request->raw[0] >> 4) & 0x30u);
#ifdef DEBUG
    fprintf(stderr, "Message type: %d\n", request->type);
#endif

    /*
     * Bit 4-7: Token Length
     */
    request->tokenLength = (request->raw[0] & 0x0fu);
#ifdef DEBUG
    fprintf(stderr, "Token length: %d\n", request->tokenLength);
#endif

    if (request->tokenLength > 8) {
        request->valid = NS_FALSE;
        return NS_FALSE;
    }

    /*
     * Bit 8-15: Token Length
     */
    code.class  = ((request->raw[1] >> 5) & 0x7u);
    code.detail = (request->raw[1] & 0x1fu);
    codeValue   = code.class * 100 + code.detail;
#ifdef DEBUG
    fprintf(stderr, "Message code: %d\n", codeValue);
#endif

    /*
     * Check, if the code belongs to the codes defined by the CoAP RFC.  If
     * not, we report the request being not valid (which might be a little
     * harsh for some applications).
     */
    for (i = 0; i < (sizeof(code_registry) / sizeof(int)); i++) {
        if (code_registry[i] == codeValue) {
            request->codeValue = codeValue;
            break;
        }
    }
    if (request->codeValue == 0) {
        return NS_FALSE;
    }

    /*
     * Bit 16-31: Message ID
     */
    request->messageID = ((unsigned int)request->raw[2] << 8) + (unsigned int)request->raw[3];
#ifdef DEBUG
    fprintf(stderr, "messageID: %d\n", request->messageID);
#endif

    /*
     * Bit 32ff: Token
     */
    request->position = 4;
    if (request->tokenLength > 0) {
        if (CheckRemainingSize(request, request->tokenLength) == NS_TRUE) {
            request->token = &(request->raw[4]);
            request->position += request->tokenLength;
#ifdef DEBUG
            fprintf(stderr, "Valid token found.\n");
#endif
        } else {
            return NS_FALSE;
        }
    }

    /*
     * Options. Option Numbers are maintained in the "CoAP Option Numbers"
     * registry.
     */
    processOptions = CheckRemainingSize(request, 1);

    while (processOptions) {
        /*
         * Option Bits:
         *    0-3: Option Delta
         *    4-7: Option Length
         *    8-15: Option Delta extended (8 bit,  when delta is 0x0d)
         *    8-23: Option Delta extended (16 bit, when delta is 0x0e)
         *    Option Length extended (8 bit, when length is 0x0d; 16 bit, when length is 0x0e)
         *    Option Value
         */
        option.delta = ((request->raw[request->position] >> 4) & 0x0fu);
        option.length = (request->raw[request->position] & 0x0fu);
        request->position++;
#ifdef DEBUG
        fprintf(stderr, "Processing option: number delta = %u, length = %u.\n",
               option.delta, option.length);
#endif
        /*
         * Parse option delta
         */
        switch (option.delta) {
            
        case 0x0fu:
            /*
             * Payload marker or invalid
             */
            switch (option.length) {
            case 0x0fu:
                request->payload = &(request->raw[request->position]);
                request->payloadLength = request->size - request->position;
#ifdef DEBUG
                fprintf(stderr, "Payload marker detected. Payload length = %d.\n",
                        request->payloadLength);
#endif
                break;
            default:
                request->valid = NS_FALSE;
                break;
            }
            processOptions = 0;
            break;

        case 0x0eu:
            if (CheckRemainingSize(request, 2) == NS_TRUE) {
                option.delta =
                    ((unsigned int)request->raw[request->position] << 8) +
                    ((unsigned int)request->raw[request->position + 1] - 269);
                request->position += 2;
            }
            break;
            
        case 0x0du:
            if (CheckRemainingSize(request, 1) == NS_TRUE) {
                option.delta = ((unsigned int)request->raw[request->position] - 13);
                request->position += 1;
            }
            break;
            
        default:
            break;
        }

        /*
         * No payload, process length
         */
        if (processOptions == 1) {
            switch (option.length) {
                
            case 0x0fu:
                request->valid = NS_FALSE;
                processOptions = 0;
                break;
                
            case 0x0eu:
                if (CheckRemainingSize(request, 2) == NS_TRUE) {
                    option.length =
                        ((unsigned int)request->raw[request->position] << 8) +
                        ((unsigned int)request->raw[request->position + 1] - 269);
                    request->position += 2;
                }
                break;
                
            case 0x0du:
                if (CheckRemainingSize(request, 1) == NS_TRUE) {
                    option.length = ((unsigned int)request->raw[request->position] - 13);
                    request->position += 1;
                }
                break;
                
            default:
                break;
            }
        }
        
        option.delta += lastOptionNumber;
        lastOptionNumber = option.delta;
#ifdef DEBUG
        fprintf(stderr, "Final option number = %u, length = %u.\n",
                option.delta, option.length);
#endif

        if (processOptions == 1) {
            if (option.length > 0) {
#ifdef DEBUG
                fprintf(stderr, "Option value expected … ");
#endif
                if (CheckRemainingSize(request, option.length) == NS_TRUE) {
#ifdef DEBUG
                    fprintf(stderr, "found.\n");
#endif
                    option.value = &(request->raw[request->position]);
                    request->position += option.length;
                } else {
#ifdef DEBUG
                    fprintf(stderr, "NOT found.\n");
#endif
                    request->valid = NS_FALSE;
                    processOptions = 0;
                }
            }
        }
        if (processOptions == 1) {
            /*
             * Append option to collection
             */
            request->options[request->optionCount] = option;
            request->optionCount++;
#ifdef DEBUG
            fprintf(stderr, "Added option to collection.\n");
#endif

            if (CheckRemainingSize(request, 1) == NS_FALSE) {
#ifdef DEBUG
                fprintf(stderr, "No further options/payload.\n");
#endif
                processOptions = 0;
            }
        }
#ifdef DEBUG
        fprintf(stderr, "Finished parsing option/payload.\n");
#endif

    }
#ifdef DEBUG
    fprintf(stderr, "parseRequest finished.\n");
#endif
    return request->valid;
}

/*
 * Translate CoAP parameters to HTTP
 */
static bool ConstructHttpRequest (CoapMsg_t *coap, HttpReq_t *http)
{
    int o, c;
    Option_t *urlComponents[coap->optionCount];
    /*
     * Method codes:
     *   CoAP supports the following methods which are a subset of those
     *   supported by HTTP
     */
    const char *methods[] = {
            "",
            "GET\0",
            "POST\0",
            "PUT\0",
            "DELETE\0"
    };
    http->method[0] = *(methods[coap->codeValue]);

    /*
     * The CoAP message-id and token are not needed by HTTP but are kept
     * for application logic and CoAP reply generation.
     */
    http->messageId = coap->messageID;
    http->tokenLength = coap->tokenLength;
    http->token = coap->token;

    http->payloadLength = coap->payloadLength;
    http->payload = coap->payload;

    /*
     * Process CoAP options
     */
    for (o = 0; o < coap->optionCount; ++o) {
        /*
         * URI options:
         *
         * [nr]  [name]
         *    3  URI host
         *    7  URI port
         *   11  URI path
         *   15  URI query
         */
        if (coap->options[o].delta & 0x3u) {
            if (coap->options[o].delta < 8) {
                strncat(&(http->host), coap->options[o].value, coap->options[o].length);
            } else if (coap->options[o].delta < 12) {
                snprintf(&http->path, coap->options[o].length, "/%s", coap->options[o].value);
            }
        }
    }

}

/*
 * Local Variables:
 * mode: c
 * c-basic-offset: 4
 * fill-column: 78
 * indent-tabs-mode: nil
 * End:
 */


nscoap-config.tcl

#
# Set the IP-address and port, on which the server listens. Since we
# want this script to run in IPv4 and IPv6 environments (and in IPv6
# environments, where IPv6 is deactivated) independent of the OS, we
# probe the interfaces here before we set the final IP address.
#
set port 8080
set address "0.0.0.0"
set address localhost

if {0 && [ns_info ipv6]} {
    #
    # The version of NaviServer supports IPv6. Probe if we can reverse
    # lookup the loopback interface and bind to the IPv6 loopback
    # address with the port specified above.
    #
    if {
	![catch {ns_hostbyaddr ::1}]
	&& ![catch {close [ns_socklisten ::1 $port]}]
    } {
	#
	# Yes we can. So use the IPv6 any address
	#
	set address ::
    }
}

#set             home                /usr/local/ns
set             home                [file dirname [file dirname [info nameofexecutable]]]

ns_section     "ns/server/default/modules"
ns_param        nscp                nscp.so
ns_param        nssock              nssock.so
ns_param        nslog               nslog.so
ns_param        nscgi               nscgi.so
ns_param        nsdb                nsdb.so
ns_param        nscoap              nscoap.so

ns_section     "ns/parameters"
ns_param        home                $home
ns_param        tcllibrary          tcl
#ns_param        tclinitlock         true	       ;# default: false
ns_param        serverlog           error.log
ns_param        pidfile             nsd.pid
#ns_param       logdebug            true               ;# default: false
#ns_param       logroll             false              ;# default: true
#ns_param	logrollfmt	    %Y-%m-%d           ;# format appended to log filename
#ns_param       dbcloseonexit       off                ;# default: off; from nsdb
ns_param        jobsperthread       1000               ;# default: 0
ns_param        jobtimeout          0                  ;# default: 300
ns_param        schedsperthread     10                 ;# default: 0
ns_param        progressminsize     [expr {1024*1024*1}] ;# default: 0
#ns_param       concurrentinterpcreate true            ;# default: false
#ns_param       listenbacklog        256               ;# default: 32; backlog for ns_socket commands
#ns_param       mutexlocktrace       true              ;# default false; print duractions of long mutex calls to stderr

# configure SMTP module
ns_param        smtphost            "localhost"
ns_param        smtpport            25
ns_param        smtptimeout         60
ns_param        smtplogmode         false
ns_param        smtpmsgid           false
ns_param        smtpmsgidhostname   ""
ns_param        smtpencodingmode    false
ns_param        smtpencoding        "utf-8"
ns_param        smtpauthmode        ""
ns_param        smtpauthuser        ""
ns_param        smtpauthpassword    ""

ns_section     "ns/threads"
ns_param        stacksize           [expr {512*1024}]

ns_section     "ns/mimetypes"
ns_param        default             text/plain
ns_param        noextension         text/plain

ns_section     "ns/db/drivers"
#ns_param       postgres           nsdbpg.so

ns_section     "ns/db/pools"
#ns_param       postgres           "PostgreSQL Database"

ns_section     "ns/db/pool/pgsql"
ns_param        driver              postgres
ns_param        connections         64
ns_param        user                postgres
ns_param        datasource          "::dbname"
ns_param        verbose             Off
ns_param        logsqlerrors        On
ns_param        extendedtableinfo   On
ns_param        maxidle             31536000
ns_param        maxopen             31536000

ns_section     "ns/servers"
ns_param        default             "NaviServer"

ns_section     "ns/server/default"
ns_param        enabletclpages      true  ;# default: false
ns_param        checkmodifiedsince  false ;# default: true, check modified-since before returning files from cache. Disable for speedup
ns_param        connsperthread      1000  ;# default: 0; number of connections (requests) handled per thread
ns_param        minthreads          5     ;# default: 1; minimal number of connection threads
ns_param        maxthreads          100   ;# default: 10; maximal number of connection threads
ns_param        maxconnections      100   ;# default: 100; number of allocated connection structures
ns_param        threadtimeout       120   ;# default: 120; timeout for idle theads
#ns_param concurrentcreatethreshold 100   ;# default: 80; allow concrruent creates when queue is fully beyond this percentage
                                          ;# 100 is a conservative value, disabling concurrent creates

ns_section     "ns/server/default/db"
ns_param        pools               *

ns_section      "ns/fastpath"
ns_param        cache               false     ;# default: false
ns_param        cachemaxsize        10240000  ;# default: 1024*10000
ns_param        cachemaxentry       8192      ;# default: 8192
ns_param        mmap                false     ;# default: false

ns_section     "ns/server/default/fastpath"
ns_param        pagedir             pages
#ns_param       serverdir           ""
ns_param        directoryfile       "index.adp index.tcl index.html index.htm"
ns_param        directoryproc       _ns_dirlist
ns_param        directorylisting    fancy    ;# default: simple
#ns_param        directoryadp       dir.adp

ns_section     "ns/server/default/vhost"
ns_param        enabled             false
ns_param        hostprefix          ""
ns_param        hosthashlevel       0
ns_param        stripport           true
ns_param        stripwww            true

ns_section     "ns/server/default/adp"
ns_param        map                 "/*.adp"
ns_param        enableexpire        false    ;# default: false; set "Expires: now" on all ADP's 
#ns_param        enabledebug         true    ;# default: false
ns_param        singlescript        false    ;# default: false; collapse Tcl blocks to a single Tcl script
ns_param        cache               false    ;# default: false; enable ADP caching
ns_param        cachesize           [expr {5000*1024}]

ns_section     "ns/server/default/tcl" {
    ns_param    initcmds {
	ns_register_proc POST /foo/bar {
	    set content [ns_getcontent -as_file false -binary false]
	    ns_return 200 text/plain POST-okidoki([string length $content])
	}
	ns_register_proc GET /foo/bar {
	    ns_return 200 text/plain GET-okidoki
	}
    }
}

ns_section     "ns/server/default/module/nscgi"
ns_param        map                 "GET  /cgi-bin $home/cgi-bin"
ns_param        map                 "POST /cgi-bin $home/cgi-bin"
ns_param        interps              CGIinterps
#ns_param        allowstaticresources true    ;# default false; allow one to serve static resources from cgi directories

ns_section "ns/interps/CGIinterps"
ns_param	.pl		    "/opt/local/bin/perl"
ns_param	.sh		    "/bin/bash"

ns_section     "ns/server/default/module/nslog"
ns_param        file                access.log
ns_param        rolllog             true     ;# default: true; should server log files automatically
ns_param        rollonsignal        false    ;# default: false; perform roll on a sighup
ns_param        rollhour            0        ;# default: 0; specify at which hour to roll
ns_param        maxbackup           7        ;# default: 10; max number of backup log files
#ns_param       rollfmt		    %Y-%m-%d-%H:%M	;# format appended to log filename
#ns_param       logpartialtimes     true     ;# default: false
#ns_param	logreqtime	    true     ;# default: false; include time to service the request


ns_section     "ns/server/default/module/nssock"
ns_param        port                 $port
#ns_param        address             0.0.0.0
#ns_param        address             ::0   ;# ::1 corresponds to 127.0.0.1, ::0 is the "unspecified address"
ns_param        address             $address
ns_param        hostname            [ns_info hostname]
ns_param        maxinput            [expr {1024*1024*10}] ;# default: 1024*1024, maximum size for inputs (uploads)
#ns_param        readahead           [expr {1024*1024*1}]  ;# default: 16384; size of readahead for requests
ns_param        backlog             1024         ;# default: 256; backlog for listen operations
ns_param        acceptsize          10           ;# default: value of "backlog"; max number of accepted (but unqueued) requests
ns_param        closewait           0            ;# default: 2; timeout in seconds for close on socket
ns_param        maxqueuesize        1024         ;# default: 1024; maximum size of the queue
ns_param        keepwait	    5	         ;# 5, timeout in seconds for keep-alive
ns_param        keepalivemaxuploadsize	 500000	 ;# 0, don't allow keep-alive for upload content larger than this
ns_param        keepalivemaxdownloadsize 1000000 ;# 0, don't allow keep-alive for download content larger than this
#
# Spooling Threads
#
#ns_param       spoolerthreads	1	;# default: 0; number of upload spooler threads
ns_param       maxupload	1000000 ;# default: 0, when specified, spool uploads larger than this value to a temp file
ns_param       writerthreads	1	;# default: 0, number of writer threads
#ns_param       writersize	1048576	;# default: 1024*1024, use writer threads for files larger than this value
#ns_param       writerbufsize	8192	;# default: 8192, buffer size for writer threads
#ns_param       driverthreads   2	;# default: 1, number of driver threads (requires support of SO_REUSEPORT)


ns_section     "ns/server/default/module/nscp"
ns_param        port                4080
#ns_param        address             127.0.0.1
ns_param        address             $address

ns_section     "ns/server/default/module/nscp/users"
ns_param        user                "::"

ns_section     "ns/server/default/module/nscoap"
ns_param        address              $address
ns_param        port                5683
ns_param        mapHTTP             "GET /foo/*"
ns_param        mapHTTP             "POST /foo/*"

#
# For debugging, you might activate one of the following flags
#
#ns_logctl severity Debug(ns:driver) on
#ns_logctl severity Debug(request) on
#ns_logctl severity Debug(task) on
#ns_logctl severity Debug(sql) on
ns_logctl severity Debug(coap) on
#ns_logctl severity Debug on

######################################################################
# One can send the actual value of an nsv array directly as a reply to
# a coap request. To do so, set the nsv "nscoap" for the interesting
# keys to some values. These values can be set by other requests,
# sensors, etc. To set such values for testing purposes, add lines
# like the following to /usr/local/ns/tcl/nscoap.tcl
#
# nsv_set nscoap hello 123
# nsv_set nscoap core foo
